lib/workflow/task-graph.js

```diff
@@ -0,0 +1,493 @@
+// Copyright 2015, EMC, Inc.
+'use strict';
+
+module.exports = taskGraphFactory;
+taskGraphFactory.$provide = 'TaskGraph.TaskGraph';
+taskGraphFactory.$inject = [
+    'Task.taskLibrary',
+    'TaskGraph.Store',
+    'Constants',
+    'Assert',
+    'uuid',
+    'Promise',
+    '_'
+];
+
+function taskGraphFactory(
+    taskLibrary,
+    store,
+    Constants,
+    assert,
+    uuid,
+    Promise,
+    _
+) {
+    function TaskGraph(definition, context, domain) {
+        this.definition = definition;
+
+        if (this.definition.options && this.definition.options.instanceId) {
+            this.instanceId = this.definition.options.instanceId;
+        } else {
+            this.instanceId = uuid.v4();
+        }
+
+        // Bool
+        this.serviceGraph = this.definition.serviceGraph;
+        this.context = context || {};
+        this.context.graphId = this.instanceId;
+        this.domain = domain || Constants.DefaultTaskDomain;
+        this.name = this.definition.friendlyName;
+        this.injectableName = this.definition.injectableName;
+
+        this.tasks = {};
+
+        // TODO: find instances of 'valid' elsewhere and replace from valid to Pending
+        // TODO: replace _status with status
+        this._status = Constants.TaskStates.Pending;
```

Comment:
It's a terrible variable name that I chose like a year ago. I'd like to change it, but I also don't want to break backwards compatibility. My thought is we could add another value "state" and start using both, then eventually deprecate "_status".